contrib/core-contract-tests/tests/pox-4/pox-4.prop.test.ts

```diff
@@ -0,0 +1,1605 @@
+import {
+  Cl,
+  ClarityType,
+  ClarityValue,
+  isClarityType,
+  serializeCV,
+} from "@stacks/transactions";
+import fc from "fast-check";
+import { assert, describe, expect, it } from "vitest";
+import { createHash } from "crypto";
+
+// Contract Consts
+const INITIAL_TOTAL_LIQ_SUPPLY = 1_000_000_000_000_000;
+const MIN_AMOUNT_USTX = 125_000_000_000n;
+const TESTNET_PREPARE_CYCLE_LENGTH = 50;
+const TESTNET_REWARD_CYCLE_LENGTH = 1050;
+const TESTNET_STACKING_THRESHOLD_25 = 8000;
+// Clarity
+const MAX_CLAR_UINT = 340282366920938463463374607431768211455n;
+const TESTNET_CHAIN_ID = 2147483648;
+const SIP_018_MESSAGE_PREFIX = "534950303138";
+// Error Codes
+const ERR_STACKING_INVALID_LOCK_PERIOD = 2;
+const ERR_STACKING_THRESHOLD_NOT_MET = 11;
+const ERR_STACKING_INVALID_POX_ADDRESS = 13;
+const ERR_STACKING_INVALID_AMOUNT = 18;
+
+function sha256(data: Buffer): Buffer {
+  return createHash("sha256").update(data).digest();
+}
+
+function structuredDataHash(structuredData: ClarityValue): Buffer {
+  return sha256(Buffer.from(serializeCV(structuredData)));
+}
+
+const generateDomainHash = (): ClarityValue =>
+  Cl.tuple({
+    name: Cl.stringAscii("pox-4-signer"),
+    version: Cl.stringAscii("1.0.0"),
+    "chain-id": Cl.uint(TESTNET_CHAIN_ID),
+  });
+
+const generateMessageHash = (
+  version: number,
+  hashbytes: number[],
+  reward_cycle: number,
+  topic: string,
+  period: number,
+  auth_id: number,
+  max_amount: number
+): ClarityValue =>
+  Cl.tuple({
+    "pox-addr": Cl.tuple({
+      version: Cl.buffer(Uint8Array.from([version])),
+      hashbytes: Cl.buffer(Uint8Array.from(hashbytes)),
+    }),
+    "reward-cycle": Cl.uint(reward_cycle),
+    topic: Cl.stringAscii(topic),
+    period: Cl.uint(period),
+    "auth-id": Cl.uint(auth_id),
+    "max-amount": Cl.uint(max_amount),
+  });
+
+const generateMessagePrefixBuffer = (prefix: string) =>
+  Buffer.from(prefix, "hex");
+
+export const buildSignerKeyMessageHash = (
+  version: number,
+  hashbytes: number[],
+  reward_cycle: number,
+  topic: string,
+  period: number,
+  max_amount: number,
+  auth_id: number
+) => {
+  const domain_hash = structuredDataHash(generateDomainHash());
+  const message_hash = structuredDataHash(
+    generateMessageHash(
+      version,
+      hashbytes,
+      reward_cycle,
+      topic,
+      period,
+      auth_id,
+      max_amount
+    )
+  );
+  const structuredDataPrefix = generateMessagePrefixBuffer(
+    SIP_018_MESSAGE_PREFIX
+  );
+
+  const signer_key_message_hash = sha256(
+    Buffer.concat([structuredDataPrefix, domain_hash, message_hash])
+  );
+
+  return signer_key_message_hash;
+};
+
+describe("test pox-4 contract", () => {
```

Comment:
vitest reports the name of the file so this `describe` isn't needed. If we remove it we'll save 2 spaces of indentation.